src/components/utils/src/timer.cc

```diff
@@ -173,46 +135,137 @@ void timer::Timer::OnTimeout() {
   LOG4CXX_DEBUG(logger_,
                 "Timer has finished counting. Timeout(ms): "
                     << static_cast<uint32_t>(timeout_ms_));
-  {
-    // Task locked by own lock because from this task in callback we can
-    // call Stop of this timer and get DeadLock
-    sync_primitives::AutoLock auto_lock(task_lock_);
-    DCHECK(task_);
-    task_->run();
-  }
+
+  // Task locked by own lock because from this task in callback we can
+  // call Stop of this timer and get DeadLock
+  sync_primitives::AutoLock auto_lock(task_lock_);
+  DCHECK(task_);
+  task_->run();
+}
+
+timer::Milliseconds timer::Timer::GetTimeout() const {
   sync_primitives::AutoLock auto_lock(lock_);
-  if (is_running_) {
-    const bool stop_result = StopUnsafe();
-    DCHECK_OR_RETURN_VOID(stop_result);
+  return timeout_ms_;
+}
+
+timer::Timer::TimerDelegate::TimerDelegate(Timer* timer)
```

Comment:
Why you don't like pointers? They are so cute :+1:  
http://coub.com/view/bhf28